overlord/hookstate/hookmgr.go

```diff
@@ -143,14 +143,37 @@ func (m *HookManager) Stop() {
 	m.runner.Stop()
 }
 
+func (m *HookManager) snapContext(contextID string) (context *Context, err error) {
+	var contexts map[string]string
+	err = m.state.Get("snap-contexts", &contexts)
+	if err != nil {
+		if err != state.ErrNoState {
+			return nil, fmt.Errorf("failed to get snap contexts: %v", err)
+		}
+	}
+	if snapName, ok := contexts[contextID]; ok {
+		// create new ephemeral context
+		context, err = NewContext(nil, m.state, &HookSetup{Snap: snapName}, nil, contextID)
+		return context, err
+	}
+	return nil, fmt.Errorf("unknown snap context requested")
+}
+
 // Context obtains the context for the given context ID.
 func (m *HookManager) Context(contextID string) (*Context, error) {
 	m.contextsMutex.RLock()
 	defer m.contextsMutex.RUnlock()
 
+	var err error
 	context, ok := m.contexts[contextID]
 	if !ok {
-		return nil, fmt.Errorf("no context for ID: %q", contextID)
+		m.state.Lock()
```

Comment:
A bit strange to see lock happening here, as it assumes an understanding of the internals of the function. Should be done inside ephemeralContext itself I think.